src/Workspaces/Core/Portable/Shared/Extensions/SemanticModelExtensions.cs

```diff
@@ -171,5 +174,62 @@ public static SemanticMap GetSemanticMap(this SemanticModel semanticModel, Synta
 
             return new TokenSemanticInfo(declaredSymbol, aliasSymbol, allSymbols, type, convertedType, token.Span);
         }
+
+        public static string GenerateNameFromType(this SemanticModel semanticModel, ITypeSymbol type, ISyntaxFacts syntaxFacts, bool capitalize)
+        {
+            var pluralize = semanticModel.Pluralize(type);
+            var typeArguments = type.GetAllTypeArguments();
+
+            // We may be able to use the type's arguments to generate a name if we're working with an enumerable type.
+            if (pluralize && TryGenerateNameFromTypeArgument(syntaxFacts, typeArguments, capitalize, out var typeArgumentParameterName))
+            {
+                return typeArgumentParameterName;
+            }
+
+            // If there's no type argument and we have an array type, we should pluralize, e.g. frogs[] instead of frog[]
+            if (type.TypeKind == TypeKind.Array && typeArguments.IsEmpty)
```

Comment:
if you have an array, we should probaly use .ElementType to drive the plural name.

Reply:
Would that not be the same behavior as the current code?
I think using ElementType would require casting to ArrayTypeSymbol, right? But this class is language agnostic and so I don't think we can make the cast here.

Reply:
> Would that not be the same behavior as the current code?

i'm not sure.  would it be? :)

> I think using ElementType would require casting to ArrayTypeSymbol, right?

yup!

> But this class is language agnostic 

So the IXXXSymbols are already language agnostic.  it's totally fine to do something like `if (type is IArrayTypeSymbol arrayType ...)`

Reply:
> i'm not sure. would it be? :)

I don't *think* so? The check verifies that we have an array with no type arguments, so I believe that leaves us only with cases like `new Frog[]`. The existing code will generate `frogs` which I think is the best name.
For arrays with a type argument, the element type is already used in an earlier check.

Reply:
> For arrays with a type argument,

FWIW, this cannot happen :)

You can think of 'arrays' as a way that generics worked *prior* to generics existing.  But if arrays are handled to already pluralize elsewhere, i guess we don't need to do anything special here :)

Reply:
Ah, I guess I was thinking of something like `Array<Integer>`, but I guess that would fall under a class with a type argument instead of an array with a type argument?

I think we do need to keep in the check to handle the `int[]` array case, but I believe the current logic is equivalent to using `ElementType`.

Reply:
> Ah, I guess I was thinking of something like Array<Integer>, but I guess that would fall under a class with a type argument instead of an array with a type argument?

Correct :)

If we had generics in .net 1.0 there likely wouldn't have been this ultra special 'array' type.  it would have likely been, as you said: `Array<T>`.  but System.Array dates back to 1.0, non-generic, days and is very special in how this type information is encoded (both for the runtime and language) :)

>but I believe the current logic is equivalent to using ElementType.

SGTM as long as there's a test to validate.